operator/pkg/lbipam/lbipam.go

```diff
@@ -0,0 +1,1728 @@
+		ipv6Enabled:       option.Config.IPv6Enabled(),
+	}
+
+	// Make a new Ctx which we will use to manage the lifetime of LBIPAM.Run.
+	ipamCtx, cancel := context.WithCancel(context.Background())
+	var wg sync.WaitGroup
+
+	params.LC.Append(hive.Hook{
+		OnStart: func(ctx hive.HookContext) error {
+			params.Logger.Info("Starting LB IPAM")
+
+			wg.Add(1)
+
+			// Start go routine, we can't block in OnStart
+			go func() {
+				defer wg.Done()
+				lbIPAM.Run(ipamCtx)
+			}()
+
+			return nil
+		},
+		OnStop: func(ctx hive.HookContext) error {
+			// Close the ipamCtx
+			cancel()
+
+			// Convert blocking Wait into a channel so we can select on the context closing or the Wait.
+			wgDone := make(chan struct{})
+			go func() {
+				wg.Wait()
+				close(wgDone)
+			}()
+
+			select {
+			case <-wgDone:
+			case <-ctx.Done():
+			}
+
+			return nil
+		},
+	})
+
+	return lbIPAM
+}
+
+// LBIPAM is the loadbalancer IP address manager, watcher/controller which allocates and assigns IP addresses
+// to LoadBalancer services from the configured set of LoadBalancerIPPools in the cluster.
+type LBIPAM struct {
+	logger logrus.FieldLogger
+
+	lbClasses   []string
+	ipv4Enabled bool
+	ipv6Enabled bool
+
+	poolClient cilium_client_v2alpha1.CiliumLoadBalancerIPPoolInterface
+	svcClient  client_typed_v1.ServicesGetter
+
+	poolResource resource.Resource[*cilium_api_v2alpha1.CiliumLoadBalancerIPPool]
+	svcResource  resource.Resource[*core_v1.Service]
+
+	poolStore    poolStore
+	rangesStore  rangesStore
+	serviceStore serviceStore
+
+	ctx context.Context
+
+	// Only used during testing.
+	initDoneCallbacks []LBIPAMInitDone
+}
+
+type ipPoolEvent = resource.Event[*cilium_api_v2alpha1.CiliumLoadBalancerIPPool]
+type svcEvent = resource.Event[*core_v1.Service]
+
+func (ipam *LBIPAM) Run(ctx context.Context) {
+	ipam.ctx = ctx
+
+	errChan := make(chan error, 2)
+
+	poolChan := stream.ToChannel[ipPoolEvent](ctx, errChan, ipam.poolResource)
+	svcChan := stream.ToChannel[svcEvent](ctx, errChan, ipam.svcResource)
+
+	err := ipam.init(ctx, poolChan, svcChan)
+	if err != nil {
+		ipam.logger.WithError(err).Error("Error while initializing LB-IPAM")
+		// Stop, don't go into main control loop
+		return
+	}
+
+	for _, cb := range ipam.initDoneCallbacks {
+		if cb != nil {
+			cb()
+		}
+	}
+
+	// When the ctx expires, both streams will close, keep processing until both channels are closed to avoid blocking
+	// the resources in case they are re-used after we shutdown.
+	for poolChan != nil && svcChan != nil {
+		select {
+		case event, ok := <-poolChan:
+			if !ok {
+				poolChan = nil
+				continue
+			}
+
+			event.Handle(
+				nil,
+				ipam.poolOnUpsert,
+				ipam.poolOnDelete,
+			)
+		case event, ok := <-svcChan:
+			if !ok {
+				svcChan = nil
+				continue
+			}
+
+			event.Handle(
+				nil,
+				ipam.svcOnUpsert,
+				ipam.svcOnDelete,
+			)
+		case err := <-errChan:
+			ipam.logger.WithError(err).Error("Resource error")
```

Comment:
We should do more than just log this error. The handlers may choose to retry and we may run out of retries, in which case we must either `Shutdown()`, or restart from clean state, but we should do something and not just leave things hanging.